### PR TITLE
Implement algorithm for calculating species tree topology distributions.

### PR DESCRIPTION
**Pull request by daniel-goldstein**
This implements both per-tree and incremental algorithms for computing exact tree topology distributions on tree sequences.

As discussed in my talk last week, this uses a DP approach to compute all possible species topology distributions at each node in the tree, starting from leaves up to roots. The `TopologyTree` class is used to store topology distributions at each node. This approach does not support internal samples, and enforces that samples be a subset of leaves (leaves that are not samples are ignored). This does, nicely, support multiply-rooted trees. As there can't be species topologies that span different roots, we compute the topologies for each root using `TopologyTree` and "add" the resulting distributions together.

Since this approach needs only a single upward traversal, the incremental algorithm is straightforward. We maintain a reference to the `TopologyTree` for each node and reconstruct those `TopologyTree`s on the upward traversals when inserting/removing edge diffs.

Should we follow the pattern of adding wrapper methods on `Tree` and `TreeSequence` (and move the docstrings over) to not have to expose the combinatorics module? If so, should we lift the functions over entirely or make the `Tree{Sequence}` methods one-liners?

cc @jeromekelleher @hyanwong

### CODE REVIEW

File: python/tests/test_combinatorics.py [hunk 1]
```diff
@@ -23,8 +23,13 @@
 """
 Test cases for combinatorial algorithms.
 """
+import io
 import itertools
 import unittest
+from collections import Counter
```
Comment (jeromekelleher) [inline python/tests/test_combinatorics.py:29]: I prefer to just import the module and then use ``collections.Counter`` in the code. This makes it obvious where the ``Counter`` class comes from. Otherwise, you see a ``Counter`` class in the code and you're not quite sure if it's the standard library's one, or a home grown version.

File: python/tskit/combinatorics.py [hunk 2]
```diff
@@ -26,11 +26,221 @@
...
+        key = get_population
+
+    topology_tree = np.full(ts.num_nodes, None, dtype=object)
+    parent = np.full(ts.num_nodes, -1)
+
+    def update_state(tree, u):
+        stack = [u]
+        while len(stack) > 0:
+            v = stack.pop()
+            children = []
```
Comment (jeromekelleher) [inline python/tskit/combinatorics.py:66]: Might as well use the high-level tree way here:
```python
children = tree.children(v)
for c in children:
     # etc
```

File: python/tskit/trees.py [hunk 3]
```diff
@@ -833,6 +833,37 @@ def unrank(rank, num_leaves):
...
+        Internal samples are not supported.
+
+        Produces a dictionary from a ``frozenset`` of labels to
+        ``collections.Counter`` of topologies, where each tree topology is represnted
+        by its rank (see :ref:`sec_combinatorics`). This represents the
+        topologies for a particular set of labels and the number of times
+        each topology appears when subsampling samples from ``tree``.
+
+        :param tree tskit.Tree: The tree to compute the topologies for.
+        :param key: A function of the form ``key(u)`` where ``u`` is a
```
Comment (jeromekelleher) [inline python/tskit/trees.py:852]: Let's change ``key`` to ``sample_sets`` so that we're using the same concepts as the stats API.

File: python/tests/test_combinatorics.py [hunk 4]
```diff
@@ -413,3 +425,302 @@ def test_is_symmetrical(self):
         self.assertFalse(three_leaf_asym.is_symmetrical())
         six_leaf_sym = RankTree(children=[three_leaf_asym, three_leaf_asym])
         self.assertTrue(six_leaf_sym.is_symmetrical())
+
+
+class TestCountTopologies(unittest.TestCase):
+    def verify_topologies(self, ts, expected=None):
+        populations = [pop.id for pop in ts.populations()]
+        topologies = [t.count_topologies() for t in ts.trees()]
+        inc_topologies = list(ts.count_topologies())
```
Comment (daniel-goldstein) [inline python/tests/test_combinatorics.py:434]: Comment here explaining subsets of populations topologies

**Comment by daniel-goldstein** [comment]
@jeromekelleher Addressed the changes we discussed. Created a `TopologyCounter` which is indexable by a tuple of sample set indexes that's returned from `count_topologies`, and a `PartialTopologyCounter` that handles most of the heavy lifting in the algorithm. With these in place there wasn't much point to a `TopologyTree`, so the per-node state in the counting topologies methods are `TopologyCounter`s.

File: python/tskit/trees.py [hunk 5]
```diff
@@ -833,6 +833,55 @@ def unrank(rank, num_leaves):
...
+        :param list sample_sets: A list of lists of Node IDs, specifying the
+            groups of individuals to compute the statistic with.
+            Defaults to all samples grouped by population.
+        :rtype: tskit.TopologyCounter
+        :raises ValueError: If nodes in ``sample_sets`` are invalid or are
+            internal samples.
+        """
+        if sample_sets is None:
+            sample_sets = [
+                list(self.tree_sequence.samples(population=pop.id))
```
Comment (jeromekelleher) [inline python/tskit/trees.py:879]: Seems a waste to convert to list here - a numpy array should be fine, right?

File: python/tests/test_combinatorics.py [hunk 6]
```diff
@@ -420,3 +433,331 @@ def test_is_symmetrical(self):
...
+        ]
+
+        pop_configurations = [
+            msprime.PopulationConfiguration(sample_size=size) for size in sample_sizes
+        ]
+        return msprime.simulate(
+            population_configurations=pop_configurations,
+            migration_matrix=migration_matrix,
+            recombination_rate=0.1,
+        )
```
Comment (jeromekelleher) [inline python/tests/test_combinatorics.py:763]: As there's a lot going on in the PartialTopologyCounter, it'd be nice to see a few simple tests for it here. Nothing exhaustive, just a handful of simple examples demonstrating that it does what it's supposed to.

Comment (daniel-goldstein) [reply python/tests/test_combinatorics.py:763]: Definitely, I think the tests definitely make it easier to pick apart.

File: python/tests/test_combinatorics.py [hunk 7]
```diff
@@ -420,3 +433,469 @@ def test_is_symmetrical(self):
...
+        with self.assertRaises(IndexError):
+            list(ts.count_topologies(sample_sets))
+
+    def test_standard_msprime_migrations(self):
+        for num_populations in range(2, 5):
+            samples = [5] * num_populations
+            ts = self.simulate_multiple_populations(samples)
+            self.verify_topologies(ts)
+
+    def test_complex_dtwf_topologies(self):
```
Comment (jeromekelleher) [inline python/tests/test_combinatorics.py:851]: This is good, but I was thinking more about the *forwards time* trees that we get out of the local Wright-Fisher simulator. Something like [these three tests](https://github.com/tskit-dev/tskit/blob/master/python/tests/test_stats.py#L293) would be really good. We want to see, in particular, how it deals with the nasty raggedy trees that a WF simulation produces.

File: python/tskit/trees.py [hunk 8]
```diff
@@ -833,6 +833,57 @@ def unrank(rank, num_leaves):
...
+        :param list sample_sets: A list of lists of Node IDs, specifying the
+            groups of nodes to compute the statistic with.
+            Defaults to all samples grouped by population.
+        :rtype: tskit.TopologyCounter
+        :raises ValueError: If nodes in ``sample_sets`` are invalid or are
+            internal samples.
+        """
+        if sample_sets is None:
+            sample_sets = [
+                np.array(self.tree_sequence.samples(population=pop.id))
```
Comment (jeromekelleher) [inline python/tskit/trees.py:881]: We don't need this cast, do we? I'm sure tree_sequence.samples returns a numpy array.

File: python/tests/test_combinatorics.py (hunk 7 above)
Comment (daniel-goldstein) [reply python/tests/test_combinatorics.py:851]: Those tests work great just drop in. Cool also to test against all sample set partitions!

File: python/tskit/combinatorics.py [hunk 9]
```diff
@@ -24,13 +24,244 @@
...
+
+class TopologyCounter:
+    """
+    Contains the distributions of embedded topologies for every combination
+    of the sample sets used to generate the ``TopologyCounter``. It is
+    indexable by a combination of sample set indexes and returns a
+    ``collections.Counter`` whose keys are topology ranks
+    (see :ref:`sec_tree_ranks`), and whose counts represent the probability
+    of that topology if one randomly selected an sample from each
+    sample set pertaining to that ``Counter``.
```
Comment (petrelharp) [inline python/tskit/combinatorics.py:135]: the last part of this sentence doesn't quite make sense to me?

File: python/tests/test_combinatorics.py [hunk 10]
```diff
@@ -420,3 +445,516 @@ def test_is_symmetrical(self):
...
+            migration_matrix=migration_matrix,
+            population_configurations=population_configurations,
+            demographic_events=demographic_events,
+            random_seed=2,
+            model="dtwf",
+        )
+
+        self.verify_topologies(ts)
+
+    def test_forward_time_wright_fisher(self):
```
Comment (jeromekelleher) [inline python/tests/test_combinatorics.py:922]: Be a little nicer to split this up into three tests, which are named. That way we can see the different types of forward sim topology that we're testing against. Big monolithic tests are hard to parse quicky for the reader.

File: python/tskit/trees.py [hunk 11]
```diff
@@ -833,6 +833,57 @@ def unrank(rank, num_leaves):
         """
         return combinatorics.RankTree.unrank(rank, num_leaves).to_tsk_tree()
 
+    def count_topologies(self, sample_sets=None):
+        """
+        Calculates the distribution of embedded topologies for every combination
+        of the sample sets in ``sample_sets``. ``sample_sets`` defaults to all
+        samples in the tree grouped by population. In this case, the topologies
+        produced represent evolutionary topologies of the populations in this
+        tree.
```
Comment (petrelharp) [inline python/tskit/trees.py:842]: I think you can delete this last sentence - I don't think it's adding anything. If you keep it, don't use "evolutionary" - there's nothing about evolution here - perhaps you mean "genealogical"?

File: python/tskit/combinatorics.py (hunk 9 above)
Comment (petrelharp) [reply python/tskit/combinatorics.py:135]: After reading the documentation for `count_topologies` I know what this does - maybe just refer to that method?

File: python/tskit/trees.py [hunk 12]
```diff
@@ -833,6 +833,57 @@ def unrank(rank, num_leaves):
         """
         return combinatorics.RankTree.unrank(rank, num_leaves).to_tsk_tree()
 
+    def count_topologies(self, sample_sets=None):
+        """
+        Calculates the distribution of embedded topologies for every combination
+        of the sample sets in ``sample_sets``. ``sample_sets`` defaults to all
+        samples in the tree grouped by population. In this case, the topologies
```
Comment (petrelharp) [inline python/tskit/trees.py:840]: do the sample sets have to be nonoverlapping? (this affects the math below)

File: python/tskit/trees.py [hunk 13]
```diff
@@ -833,6 +833,57 @@ def unrank(rank, num_leaves):
...
+        populations' samples in this tree.
+
+        The counts for each topology in the :class:`tskit.TopologyCounter`
+        are absolute counts that we would get if we were to select all
+        combinations of samples from the relevant sample sets.
+        For sample sets :math:`[s_0, s_1, ..., s_n]`, the total number of
+        topologies for those sample sets is equal to
+        :math:`|s_0| * |s_1| * ... * |s_n|`, so the counts in the counter
+        ``topology_counter[0, 1, ..., n]`` should sum to
+        :math:`|s_0| * |s_1| * ... * |s_n|`.
```
Comment (petrelharp) [inline python/tskit/trees.py:861]: nice job making it concrete!

File: python/tskit/trees.py [hunk 14]
```diff
@@ -5840,6 +5891,28 @@ def kc_distance(self, other, lambda_=0.0):
         """
         return self._ll_tree_sequence.get_kc_distance(other._ll_tree_sequence, lambda_)
 
+    def count_topologies(self, sample_sets=None):
+        """
+        Generates the same distribution of topologies as :meth:`Tree.count_topologies`
+        but for every tree in a tree sequence. For use on a tree sequence this method is
```
Comment (petrelharp) [inline python/tskit/trees.py:5897]: How exactly is this combined across all the trees? Is it added together by tree? Weighted by span?

**Comment by petrelharp** [comment]
Looks good! I had some clarifying comments on the docs, and am happy to look again if you want, but don't have to.

File: python/tskit/trees.py (hunk 12 above)
Comment (daniel-goldstein) [reply python/tskit/trees.py:840]: Yes they do, good point.

File: python/tskit/trees.py (hunk 14 above)
Comment (daniel-goldstein) [reply python/tskit/trees.py:5897]: They are not combined, it produces a generator that yields the per-tree result. Can clarify

**Comment by daniel-goldstein** [comment]
Thanks @petrelharp!

File: python/tskit/trees.py (hunk 13 above)
Comment (daniel-goldstein) [reply python/tskit/trees.py:861]: Thanks!

**Comment by jeromekelleher** [comment]
OK, great, I'm going to mark this for merging.

@daniel-goldstein, can you open issues to track anything we've left dangling/unfinished here please?

**Comment by benjeffery** [comment]
@mergifyio refresh

**Comment by benjeffery** [comment]
Travis AWOL again! Kicking it.
